### PR TITLE
Ryan Smith

### PR DESCRIPTION
**Pull request by RyanChrisSmith**
Had some real trouble in the beginning with the Shoulda-matcher gem. It wasnt loading to my machine. Ended up having to sudo install it, but wasted quite a bit of time solving that issue. 

I didnt get further than finishing the first story and testing 1 edge case, but would love feedback on a cleaner way to get the 'total count' from the api call.  It feels kind of code smell-y the way I solved it, not a fan. 

Any other 'cleaner code' feedback is welcome too!

### CODE REVIEW

File: app/controllers/quotes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class QuotesController < ApplicationController
+  def index
+    @quotes = QuoteFacade.search_term(params[:query])
+    json = QuoteService.search_by_input(params[:query])
+    @total = json[:totalCount]
```
Comment (epintozzi) [inline app/controllers/quotes_controller.rb:5]: I agree with your self assessment that this feels smelly. the `search_term` method in your facade class is calling the same `search_by_input` method in your service so you are hitting the same endpoint twice. I will add some thoughts about a better way to handle this down in the comments for the facade class.

File: app/facades/quote_facade.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class QuoteFacade
+
+  def self.search_term(input)
+    json = QuoteService.search_by_input(input)
+
+    @quotes = json[:results].map do |quote_data|
+      QuotePoro.new(quote_data)
+    end
+  end
```
Comment (epintozzi) [inline app/facades/quote_facade.rb:9]: In order to solve the issue of duplicate API calls, I would return either an array or a hash that includes all the data you want in the controller that comes from this one API call. That could look something like this:
```ruby
  def self.search_term(input)
    json = QuoteService.search_by_input(input)
    total = json[:totalCount]
    quotes = json[:results].map do |quote_data|
      QuotePoro.new(quote_data)
    end
    [total, quotes]
  end
```
and then in the controller, you would just update to:
```ruby
@total, @quotes = QuoteFacade.search_term(params[:query])
```

File: app/facades/quote_facade.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class QuoteFacade
+
+  def self.search_term(input)
+    json = QuoteService.search_by_input(input)
+
+    @quotes = json[:results].map do |quote_data|
```
Comment (epintozzi) [inline app/facades/quote_facade.rb:6]:
```suggestion
    quotes = json[:results].map do |quote_data|
```
I don't think this needs to be an instance variable. You can just use a local variable `quotes` here and it'll work just as well.

File: spec/poros/quote_poro_spec.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+RSpec.describe QuotePoro do
+  it 'will return only the info needed' do
+    quotes = QuoteService.search_by_input("extreme love")
+
+    quote_data = quotes[:results].first
```
Comment (epintozzi) [inline spec/poros/quote_poro_spec.rb:7]: This definitely works, but there are a couple things I don't love about the approach are using the service or facade to get the data to send to the poro. 1) If your service ever changes (ex: the name of the method), then this test breaks even though it's unrelated. 2) It's making an API call when all we really need to test this is some sample data in the format that the `QuotePoro` class is expecting. We don't need to use VCR or other mocking tools to do this. 3) If the first result from the API ever changes, this test fails. We can just make up data ourselves and do something like this instead:
```ruby
  before :each do
    @quote = QuotePoro.new({ content: "If I am not for myself, who will be for me? If I am not for others, what am I? And if not now, when?",
                         author: "Rabbi Hillel",
                         tags: ["famous-quotes", "wisdom"] })
  end
```

File: spec/poros/quote_poro_spec.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    quote = QuotePoro.new(quote_data)
+
+    expect(quote).to be_a QuotePoro
+    expect(quote.author).to be_a(String)
+    expect(quote.content).to be_a(String)
+    expect(quote.categories).to be_an(Array)
```
Comment (epintozzi) [inline spec/poros/quote_poro_spec.rb:14]: For our poro testing, we should be testing actual stored values which is another benefit of passing data directly to our poro and skipping the real API call. We will know exactly what the expected results are instead of using `String` and `Array` which isn't very specific.

File: spec/poros/quote_poro_spec.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+RSpec.describe QuotePoro do
```
Comment (epintozzi) [inline spec/poros/quote_poro_spec.rb:3]: I would also test what happens when tags are empty since the logic for handling that is in the poro. If it were to break, we have no way of knowing that if it's not being tested.

**Review by epintozzi — COMMENTED** [review]
Nice work! A few comments about how you can clean things up, but overall you are on the right track!
